graphium/graphium_cpp/spectral.cpp

```diff
@@ -0,0 +1,296 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "spectral.h"
+
+#include <algorithm>
+#include <assert.h>
+#include <cmath>
+#include <numeric>
+
+#include "features.h"
+#include <ATen/ops/linalg_eig.h>
+
+template<typename T>
+void compute_laplacian_eigendecomp_single(const uint32_t n, LaplacianData<T>& data, bool symmetric) {
+    T* matrix = data.matrix_temp.data();
+    std::unique_ptr<T[]> matrix_alloc(new T[n * n]);
+    std::copy(matrix, matrix + n * n, matrix_alloc.get());
+
+    int64_t dims[2] = { n, n };
+    at::Tensor torch_matrix = torch_tensor_from_array<T>(std::move(matrix_alloc), dims, 2, c10::ScalarType::Double);
+
+    at::Tensor eigenvalue_tensor;
+    at::Tensor eigenvector_tensor;
+    if (symmetric) {
+        // Using linalg_eigh should ensure we get all real eigenvalues and eigenvectors.
+        // Arbitrarily choose lower-triangular portion (L)
+        auto tuple = at::linalg_eigh(torch_matrix, c10::string_view("L",1));
+        eigenvalue_tensor = std::move(std::get<0>(tuple));
+        eigenvector_tensor = std::move(std::get<1>(tuple));
+    }
+    else {
+        auto tuple = at::linalg_eig(torch_matrix);
+        eigenvalue_tensor = std::move(std::get<0>(tuple));
+        eigenvector_tensor = std::move(std::get<1>(tuple));
+    }
+    assert(eigenvalue_tensor.ndimension() == 1);
+    assert(eigenvector_tensor.ndimension() == 2);
+    assert(eigenvalue_tensor.size(0) == n);
+    assert(eigenvector_tensor.size(0) == n);
+    assert(eigenvector_tensor.size(1) == n);
+
+    // Copy eigenvalues
+    data.eigenvalues_temp.resize(n);
+    if (eigenvalue_tensor.scalar_type() == c10::ScalarType::Double) {
+        const double* const eigenvalue_data = eigenvalue_tensor.data_ptr<double>();
+        for (size_t i = 0; i < n; ++i) {
+            data.eigenvalues_temp[i] = T(eigenvalue_data[i]);
+        }
+    }
+    else if (eigenvalue_tensor.scalar_type() == c10::ScalarType::ComplexDouble) {
+        // TODO: Decide what to do about legitimately complex eigenvalues.
+        // This should only occur with Normalization::INVERSE, because real, symmetric
+        // matrices have real eigenvalues.
+        // For now, just assume that they're supposed to be real and were only complex
+        // due to roundoff.
```

Comment:
We could also use this to our advantage. If `inverse` is asked, we could just compute the regular eigvecs/eigvals, then rescale the eigvecs, thus allowing us to use `eigh`

Reply:
Thanks for the explanation!  Yeah, it would be good if the code can exclusively use `eigh`, to avoid the issues I was hitting with `eig` sometimes producing complex eigenvectors with non-zero imaginary components.

Reply:
I've now implemented the inverse case in terms of `eigh` and removed the complex eigenvector and eigenvalue cases, because the documentation for eigh says that the types should be the same as the input types.  I haven't tested it, but the idea is in there.